src/Plugin/Field/FieldFormatter/SmartDocsFormatter.php

```diff
@@ -0,0 +1,157 @@
+<?php
+
+/**
+ * Copyright 2018 Google Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License version 2 as published by the
+ * Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public
+ * License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+namespace Drupal\apigee_api_catalog\Plugin\Field\FieldFormatter;
+
+use Drupal\Component\Serialization\Exception\InvalidDataTypeException;
+use Drupal\Core\Field\FieldDefinitionInterface;
+use Drupal\Core\Field\FieldItemListInterface;
+use Drupal\Core\Logger\LoggerChannelFactoryInterface;
+use Drupal\Core\Plugin\ContainerFactoryPluginInterface;
+use Drupal\Core\Serialization\Yaml;
+use Drupal\file\Plugin\Field\FieldFormatter\FileFormatterBase;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Drupal\Component\Serialization\Json;
+
+/**
+ * Plugin implementation of the SmartDocs OpenAPI spec formatter.
+ *
+ * @FieldFormatter(
+ *   id = "apigee_api_catalog_smartdocs",
+ *   label = @Translation("SmartDocs"),
+ *   description = @Translation("Formats OpenAPI specs with SmartDocs"),
+ *   field_types = {
+ *     "file"
```

Comment:
@cnovak I think this is supposed to be `file_link` instead of `file`.

Reply:
I changed this locally, which then allowed "SmartDocs" to be selected as the formatter, but it also caused a fatal error when trying to view the details page:

```
TypeError: Argument 1 passed to Drupal\Core\Field\Plugin\Field\FieldFormatter\EntityReferenceFormatterBase::getEntitiesToView() must implement interface Drupal\Core\Field\EntityReferenceFieldItemListInterface, instance of Drupal\Core\Field\FieldItemList given, called in /Users/jacine/Sites/kickstart_m10n/web/modules/contrib/apigee_api_catalog/src/Plugin/Field/FieldFormatter/SmartDocsFormatter.php on line 126 in Drupal\Core\Field\Plugin\Field\FieldFormatter\EntityReferenceFormatterBase->getEntitiesToView() (line 42 of /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Field/Plugin/Field/FieldFormatter/EntityReferenceFormatterBase.php) #0 /Users/jacine/Sites/kickstart_m10n/web/modules/contrib/apigee_api_catalog/src/Plugin/Field/FieldFormatter/SmartDocsFormatter.php(126): Drupal\Core\Field\Plugin\Field\FieldFormatter\EntityReferenceFormatterBase->getEntitiesToView(Object(Drupal\Core\Field\FieldItemList), 'en') #1 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Field/FormatterBase.php(80): Drupal\apigee_api_catalog\Plugin\Field\FieldFormatter\SmartDocsFormatter->viewElements(Object(Drupal\Core\Field\FieldItemList), 'en') #2 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Field/Plugin/Field/FieldFormatter/EntityReferenceFormatterBase.php(76): Drupal\Core\Field\FormatterBase->view(Object(Drupal\Core\Field\FieldItemList), 'en') #3 /Users/jacine/Sites/kickstart_m10n/web/modules/contrib/apigee_api_catalog/src/Plugin/Field/FieldFormatter/SmartDocsFormatter.php(98): Drupal\Core\Field\Plugin\Field\FieldFormatter\EntityReferenceFormatterBase->view(Object(Drupal\Core\Field\FieldItemList), NULL) #4 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Entity/Entity/EntityViewDisplay.php(263): Drupal\apigee_api_catalog\Plugin\Field\FieldFormatter\SmartDocsFormatter->view(Object(Drupal\Core\Field\FieldItemList), NULL) #5 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Entity/EntityViewBuilder.php(343): Drupal\Core\Entity\Entity\EntityViewDisplay->buildMultiple(Array) #6 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Entity/EntityViewBuilder.php(285): Drupal\Core\Entity\EntityViewBuilder->buildComponents(Array, Array, Array, 'full') #7 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Entity/EntityViewBuilder.php(242): Drupal\Core\Entity\EntityViewBuilder->buildMultiple(Array) #8 [internal function]: Drupal\Core\Entity\EntityViewBuilder->build(Array) #9 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Render/Renderer.php(378): call_user_func(Array, Array) #10 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Render/Renderer.php(195): Drupal\Core\Render\Renderer->doRender(Array, false) #11 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Render/MainContent/HtmlRenderer.php(226): Drupal\Core\Render\Renderer->render(Array, false) #12 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Render/Renderer.php(582): Drupal\Core\Render\MainContent\HtmlRenderer->Drupal\Core\Render\MainContent\{closure}() #13 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Render/MainContent/HtmlRenderer.php(227): Drupal\Core\Render\Renderer->executeInRenderContext(Object(Drupal\Core\Render\RenderContext), Object(Closure)) #14 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/Render/MainContent/HtmlRenderer.php(117): Drupal\Core\Render\MainContent\HtmlRenderer->prepare(Array, Object(Symfony\Component\HttpFoundation\Request), Object(Drupal\Core\Routing\CurrentRouteMatch)) #15 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/EventSubscriber/MainContentViewSubscriber.php(90): Drupal\Core\Render\MainContent\HtmlRenderer->renderResponse(Array, Object(Symfony\Component\HttpFoundation\Request), Object(Drupal\Core\Routing\CurrentRouteMatch)) #16 [internal function]: Drupal\Core\EventSubscriber\MainContentViewSubscriber->onViewRenderArray(Object(Symfony\Component\HttpKernel\Event\GetResponseForControllerResultEvent), 'kernel.view', Object(Drupal\Component\EventDispatcher\ContainerAwareEventDispatcher)) #17 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Component/EventDispatcher/ContainerAwareEventDispatcher.php(111): call_user_func(Array, Object(Symfony\Component\HttpKernel\Event\GetResponseForControllerResultEvent), 'kernel.view', Object(Drupal\Component\EventDispatcher\ContainerAwareEventDispatcher)) #18 /Users/jacine/Sites/kickstart_m10n/vendor/symfony/http-kernel/HttpKernel.php(156): Drupal\Component\EventDispatcher\ContainerAwareEventDispatcher->dispatch('kernel.view', Object(Symfony\Component\HttpKernel\Event\GetResponseForControllerResultEvent)) #19 /Users/jacine/Sites/kickstart_m10n/vendor/symfony/http-kernel/HttpKernel.php(68): Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object(Symfony\Component\HttpFoundation\Request), 1) #20 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/StackMiddleware/Session.php(57): Symfony\Component\HttpKernel\HttpKernel->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #21 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/StackMiddleware/KernelPreHandle.php(47): Drupal\Core\StackMiddleware\Session->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #22 /Users/jacine/Sites/kickstart_m10n/web/core/modules/page_cache/src/StackMiddleware/PageCache.php(106): Drupal\Core\StackMiddleware\KernelPreHandle->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #23 /Users/jacine/Sites/kickstart_m10n/web/core/modules/page_cache/src/StackMiddleware/PageCache.php(85): Drupal\page_cache\StackMiddleware\PageCache->pass(Object(Symfony\Component\HttpFoundation\Request), 1, true) #24 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/StackMiddleware/ReverseProxyMiddleware.php(47): Drupal\page_cache\StackMiddleware\PageCache->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #25 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/StackMiddleware/NegotiationMiddleware.php(52): Drupal\Core\StackMiddleware\ReverseProxyMiddleware->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #26 /Users/jacine/Sites/kickstart_m10n/vendor/stack/builder/src/Stack/StackedHttpKernel.php(23): Drupal\Core\StackMiddleware\NegotiationMiddleware->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #27 /Users/jacine/Sites/kickstart_m10n/web/core/lib/Drupal/Core/DrupalKernel.php(693): Stack\StackedHttpKernel->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #28 /Users/jacine/Sites/kickstart_m10n/web/index.php(19): Drupal\Core\DrupalKernel->handle(Object(Symfony\Component\HttpFoundation\Request)) #29 {main}.
```